src/coreclr/jit/assertionprop.cpp

```diff
@@ -3821,14 +3853,92 @@ GenTree* Compiler::optAssertionProp_LclFld(ASSERT_VALARG_TP assertions, GenTreeL
 GenTree* Compiler::optAssertionProp_Asg(ASSERT_VALARG_TP assertions, GenTreeOp* asg, Statement* stmt)
 {
     GenTree* rhs = asg->gtGetOp2();
+
+    // Try and simplify the RHS.
+    //
+    bool madeChanges = false;
     if (asg->OperIsCopyBlkOp() && varTypeIsStruct(rhs))
     {
         if (optZeroObjAssertionProp(rhs, assertions))
         {
-            return optAssertionProp_Update(asg, asg, stmt);
+            madeChanges = true;
+            rhs         = asg->gtGetOp2();
+        }
+    }
+
+    // If we're assigning a value to a lcl/field that already has
+    // that value, suppress the assignment.
+    //
+    // For now we just check for zero.
+    //
+    // In particular we want to make sure that for struct S the
+    // "redundant init" pattern
+    //
+    //   S s = new S();
+    //   s.field = 0;
+    //
+    // does not kill the zerobj assertion for s.
+    //
+    if (optLocalAssertionProp)
+    {
+        GenTreeLclVarCommon* lhsVarTree = nullptr;
+        if (asg->DefinesLocal(this, &lhsVarTree))
+        {
+            unsigned const       lhsLclNum      = lhsVarTree->GetLclNum();
+            LclVarDsc* const     lhsLclDsc      = lvaGetDesc(lhsLclNum);
+            bool const           lhsLclIsStruct = varTypeIsStruct(lhsLclDsc->TypeGet());
+            AssertionIndex const lhsIndex =
+                optLocalAssertionIsEqualOrNotEqual(O1K_LCLVAR, lhsLclNum, lhsLclIsStruct ? O2K_ZEROOBJ : O2K_CONST_INT,
+                                                   0, assertions);
+            if (lhsIndex != NO_ASSERTION_INDEX)
+            {
+                AssertionDsc* const lhsAssertion = optGetAssertion(lhsIndex);
+                if ((lhsAssertion->assertionKind == OAK_EQUAL) && (lhsAssertion->op2.u1.iconVal == 0))
+                {
+                    bool canOptimize = false;
+
+                    // LHS is zero. Is RHS a literal zero? If so we don't need the assignment.
+                    //
+                    // The latter part of the if below is a heuristic.
+                    //
+                    // If we elimiate a zero assignment for integral lclVars it can lead to
+                    // unnecessary cloning. We need to make sure `optExtractInitTestIncr`
```

Comment:
> it can lead to unnecessary cloning

Not sure if I follow this.

Reply:
Say we remove a `Vxx = 0` because assertion prop already knows `Vxx` is zero, and `Vxx` is the control variable for some loop.

With the redundant zeroing removed, `optExtractdInitTestIncr` may fail to find the initial value for `Vxx` because it just does a simple local search. When this happens, the cloner will fail to prove that the loop cloning conditions are always satisfied. and so do extra cloning that it otherwise would not do.

What's a little odd here is that we can't then later on (using more powerful VN/global AP) prove that the loop cloning conditions are indeed satisfied and clean things back up as if we'd never cloned. I didn't dig into this because knowingly enabling unnecessary cloning and relying on later phases to clean up seemed like a bad plan.